ignite/contrib/handlers/trains_logger.py

```diff
@@ -0,0 +1,578 @@
+            trains_logger.attach(trainer,
+                                 log_handler=WeightsScalarHandler(model),
+                                 event_name=Events.ITERATION_COMPLETED)
+
+
+    """
+
+    def __init__(self, *_, **kwargs):
+        try:
+            import trains
+        except ImportError:
+            raise RuntimeError(
+                "This contrib module requires trains to be installed. "
+                "You may install trains using: \n pip install trains \n"
+            )
+
+        experiment_kwargs = {
+            k: v for k, v in kwargs.items() if k not in (
+                "project_name", "task_name", "task_type", "report_freq",
+                "histogram_update_freq_multiplier", "histogram_granularity"
+            )
+        }
+
+        self.task = trains.Task.init(
+            project_name=kwargs.get("project_name"),
+            task_name=kwargs.get("task_name"),
+            task_type=kwargs.get("task_type", trains.Task.TaskTypes.training),
+            **experiment_kwargs
+        )
+
+        self.trains_logger = self.task.get_logger()
+
+        self.grad_helper = trains.binding.frameworks.tensorflow_bind.WeightsGradientHistHelper(
+            logger=self.trains_logger,
+            report_freq=kwargs.get("report_freq", 100),
+            histogram_update_freq_multiplier=kwargs.get("histogram_update_freq_multiplier", 10),
+            histogram_granularity=kwargs.get("histogram_granularity", 50),
+        )
+
+
+class TrainsSaver(DiskSaver):
+    """Handler that saves input checkpoint as Trains artifacts
+
+    Examples:
+
+        .. code-block:: python
+
+            from ignite.contrib.handlers.trains_logger import *
+            from ignite.handlers import Checkpoint
+
+            trains_logger = TrainsLogger(project_name="pytorch-ignite-integration",
+                                         task_name="cnn-mnist"
+                                         )
+
+            to_save = {"model": model}
+
+            handler = Checkpoint(to_save, TrainsSaver(trains_logger), n_saved=1,
+                                 score_function=lambda e: 123, score_name="acc",
+                                 filename_prefix="best",
+                                 global_step_transform=global_step_from_engine(trainer))
+
+            validation_evaluator.add_event_handler(Events.EVENT_COMPLETED, handler)
+
+    """
+
+    def __init__(self, logger: TrainsLogger, output_uri=None, dirname=None, *args, **kwargs):
+        if not isinstance(logger, TrainsLogger):
+            raise TypeError("logger must be an instance of TrainsLogger")
+
+        try:
+            from trains import Task
+        except ImportError:
+            raise RuntimeError(
+                "This contrib module requires trains to be installed. "
+                "You may install trains using: \n pip install trains \n"
+            )
+
+        if not dirname:
+            dirname = tempfile.mkdtemp(
```

Comment:
Does this temp dir get cleaned up somehow?

Reply:
Not explicitly, but it's in the system's temp folder, so just like any other system temporary files.
The thinking was that since we allow model checkpoint logging even if `dirname` was not provided, it would be nice for a user to be able to run her experiment (even for a quick try-out, debugging or whatever) and see model checkpoints logged, pointing to files that are actually stored on the training machine (even if she didn't bother providing a target folder yet).
The `output_uri` param in the saver (or in the logger, actually) allows the user to provide an upload target for model checkpoints, but an actual file is still required, at least until the checkpoint is uploaded (asynchronously).

Reply:
Ah, I see. Makes sense. All good - just checking. 👍